pandas/core/internals/__init__.py

```diff
@@ -86,6 +86,55 @@
 from pandas.compat import range, map, zip, u
 
 
+def try_cast_result(left, result, dtype=None):
+    """
+    Try to cast the result to the original dtype for `left`; we may have
+    roundtripped thru object in the mean-time.
+
+    Parameters
+    ----------
+    left : array-like
+    result : array-like
+    dtype : np.dtype, pd.dtype, or None (default None)
+
+    Returns
+    -------
+    maybe_casted : same type as `result`
+    """
+    if dtype is None:
+        dtype = left.dtype
+
+    if (is_integer_dtype(left) or is_bool_dtype(left) or
+            is_datetime64_dtype(left) or is_datetime64tz_dtype(left)):
+        pass
+    elif is_object_dtype(left) and lib.infer_dtype(left) == 'boolean':
+        # disguised is_bool_dtype
+        pass
```

Comment:
This is nearly cut/paste from the existing `Block._try_cast_result` implementation.  The difference is this uses e.g. `is_integer_dtype(left)` instead of `left.is_integer`.  The `is_foo` attributes map directly onto the `is_foo_dtype` checks with the exception of `is_bool`, which requires this extra object-dtype check (this is the same way `Index.is_boolean` works).